### PR TITLE
Set stride order

### PR DESCRIPTION
**Pull request by jjsjann123**
- Adding python API `stride_order(arg, stride_order)`;
- Refactored OutputRecord to use allocation domain instead of permutation op to specify output stride_order;
- Refactored `PermuteOpRecord` to `DimsOpRecord` in fusion record to handle both `permute` and `stride_order` op;
- Updating stride_order semantics: each entry states the order for current dimension. So 4d contiguous will be [3, 2, 1, 0] and 4d channels-last will be [3, 0, 2, 1].

The reason to have `stride_order` op, instead of relying on `add_output` with a stride order, is to minimize book keeping. i.e. framework might sprinkle stride_order primitive in the graph without knowing whether such op will be at a fusion  boundary.
Having an op that could update the stride_order on any TensorView makes it easier to handle this in our python API when we automatically parse framework IR.

TODO:

- [x] refactor add_output to use allocation domain for stride order
- [x] add test for set
- [ ] Currently we don't clean up the allocation domain set on intermediate tensor. We can add a clean up pass at the end of fusion record if that turns out to be necessary.

### CODE REVIEW

**Comment by jjsjann123** [comment]
!build

File: csrc/python_frontend/fusion_record.h
Comment (jacobhinkle) [inline csrc/python_frontend/fusion_record.h:635]: This is permutation-invariant I think, so will have same hash for any inputs with same dimension. Maybe just shift by position before xoring, or use `hashCombine`?

Comment (jjsjann123) [reply csrc/python_frontend/fusion_record.h:635]: ohohoh. good catch! We need to fix this for permute as well. :laughing:

File: csrc/python_frontend/fusion_record.h
Comment (jjsjann123) [inline csrc/python_frontend/fusion_record.h:667]: IIUC, setting output stride is as simple as setting allocation domain on the tensor.

Note that user could also use this to manipulate the allocation domain on intermediate tensor, would that be causing any issue? I'm imagining this would just be ignored by codegen. cc'ing @zasdfgbnm

Comment (jacobhinkle) [reply csrc/python_frontend/fusion_record.h:667]: If we want to restrict it to outputs only, could we add stride order as an optional arg to `add_output` instead?

Comment (jjsjann123) [reply csrc/python_frontend/fusion_record.h:668]: We already have the stride order as part of the API for add_output.

It's a bit tricky here how `stride_order` can be specified in framework now. i.e. framework might want to have an operator like `contiguous(stride_order)` to dictate the tensor stride. If we decide to support that op, it could ended up being in an arbitrary position in the fusion.

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
CI failure seems to be some old driver on our system.

File: csrc/python_frontend/fusion_record.h
Comment (zasdfgbnm) [reply csrc/python_frontend/fusion_record.h:668]: Should we just ignore `contiguous` and only look at stride order of output? I don't understand what this is trying to support.

Comment (jjsjann123) [reply csrc/python_frontend/fusion_record.h:668]: This is just to keep our life simpler during parsing framework graphs.

i.e. framework could specify `x.contigous(channels_last)` in arbitrary places, not necessarily on a fusion output. Even though we would only care about contiguous calls on output and safely ignore any other, having a `set` op where we can specify output strides makes it easier so we don't need extra bits for bookkeeping.

Comment (zasdfgbnm) [reply csrc/python_frontend/fusion_record.h:668]: Then does it make sense to always make `contiguous` a set op?

**Comment by jjsjann123** [comment]
!build

File: csrc/python_frontend/fusion_record.h
Comment (jjsjann123) [reply csrc/python_frontend/fusion_record.h:668]: No, we do want to specify the right stride order.

i.e. we could have something like
```
t1 = normalization(t0)
t2 = contiguous(t1, channels_last)
convolution_out = convolution(t2)
```

So partition would give a nvfuser region like
```
t1 = normalization(t0)
t2 = contiguous(t1, channels_last)
# fusion.add_output(t2)
```

if we have a `set` op with stride order argument, when we parse the graph, we can simply just parse each op as-is, without tracing whether the graph is trying to set a stride-order on output or just intermediate tensor.

**Comment by jjsjann123** [comment]
!build

**Comment by xwang233** [comment]
!build

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
I'm strangely not getting a repro on local A100 with this `To reproduce: NVFUSER_TEST_RANDOM_SEED=1695238791 NVFUSER_TEST_ATEN_RANDOM_SEED=0 nvfuser_tests --gtest_filter='MatmulSchedulerTest.EpilogueBias_CUDA'`

Trying CI again.

File: csrc/python_frontend/fusion_record.h
Comment (jjsjann123) [inline csrc/python_frontend/fusion_record.h:1489]: refactored output stride order support to use allocation domain instead, this likely will give us some regression on our support since scheduler doesn't do a good job on that yet.
But I do think this is a step towards the right direction. :crossed_fingers:

File: python_tests/test_python_frontend.py
Comment (jjsjann123) [inline python_tests/test_python_frontend.py:1455]: cc'ing @kevinstephano this was what I was talking about the updated `stride_order` semantics :laughing: Hope you don't mint it too much.

**Comment by jjsjann123** [comment]
@zasdfgbnm this one is good for a round-2 review

File: csrc/python_frontend/fusion_record.h
Comment (zasdfgbnm) [reply csrc/python_frontend/fusion_record.h:659]: O think we should merge them and uses `if constexpr` for dispatching

**Comment by zasdfgbnm** [comment]
I would love to see the codegen diff check, but it is broken right now...

File: csrc/python_frontend/fusion_record.h
Comment (jjsjann123) [reply csrc/python_frontend/fusion_record.h:659]: Sounds fair. Let me clean that up~~

**Comment by jjsjann123** [comment]
!build

File: csrc/python_frontend/fusion_record.h
Comment (jjsjann123) [inline csrc/python_frontend/fusion_record.h:535]: renamed this RecordFunctor to be `DimsOpRecord` it now handles both `permute` and `stride_order`
cc'ing @zasdfgbnm

**Comment by xwang233** [comment]
!build

**Comment by jjsjann123** [comment]
bumping for review~~

**Comment by zasdfgbnm** [comment]
!build

File: csrc/python_frontend/python_bindings.cpp
Comment (rdspring1) [inline csrc/python_frontend/python_bindings.cpp:773]: Perhaps we should change this to `set_stride_order`.

**Review by rdspring1 — APPROVED** [review]
LGTM. I think we should make the function name in python frontend, so the API is clearer for users.

File: csrc/python_frontend/python_bindings.cpp
Comment (jacobhinkle) [reply csrc/python_frontend/python_bindings.cpp:773]: How about making  the `stride_order` argument keyword-only?

Comment (rdspring1) [reply csrc/python_frontend/python_bindings.cpp:773]: Thunder has an equivalent function called `stride_order` and numpy has [numpy.lib.stride_tricks.as_strided](https://numpy.org/doc/stable/reference/generated/numpy.lib.stride_tricks.as_strided.html)

@jacobhinkle Did you mean to keep the `set` name but make `stride_order` a keyword argument?

Comment (jjsjann123) [reply csrc/python_frontend/python_bindings.cpp]: renamed it to stride_order now.

Comment (jjsjann123) [reply csrc/python_frontend/python_bindings.cpp]: soft cc'ing Jacob.

I made the signature as `stride_order(TensorView, list[int])`. so stride order is required now for the API.

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
wait.. how could this PR possibly break the sinh jit example?!

**Comment by jjsjann123** [comment]
looks like the sinh example is just broken. I'm merging this PR as-is then.

**Comment by zasdfgbnm** [comment]
> looks like the sinh example is just broken. I'm merging this PR as-is then.

Is it fixed in https://github.com/NVIDIA/Fuser/pull/941?

**Comment by jjsjann123** [comment]
> > looks like the sinh example is just broken. I'm merging this PR as-is then.
> 
> Is it fixed in #941?

I don't think I had #941 in my earlier run and saw similar breakage in one of jacob's inflight PR.

Thanks for starting the pipeline. I can wait on that to finish before push the merge button.

**Comment by jacobhinkle** [comment]
> looks like the sinh example is just broken. I'm merging this PR as-is then.

See https://github.com/NVIDIA/Fuser/pull/948. I think #941 fixed `sinh_libtorch` but this fixes `sinh_extension`.

**Comment by jjsjann123** [comment]
> > looks like the sinh example is just broken. I'm merging this PR as-is then.
> 
> See #948. I think #941 fixed `sinh_libtorch` but this fixes `sinh_extension`.

forgot there are two of those. Thanks for patching this.

**Comment by jjsjann123** [comment]
failing CI is complaining about driver version.

```
what():  The NVIDIA driver on your system is too old (found version 11040). Please update your GPU driver by downloading and installing a new version from the URL: http://www.nvidia.com/Download/index.aspx Alternatively, go to: https://pytorch.org/ to install a PyTorch version that has been compiled with your version of the CUDA driver.
Exception raised from device_count_impl at ../c10/cuda/CUDAFunctions.cpp:45 (most recent call first):
```

**Comment by jjsjann123** [comment]
Don't think there's any real failure. I'm merging this one.

**Pull request merged**
